File: json/json-annotation-processor/src/test/java/ru/tinkoff/kora/json/annotation/processor/JsonIncludeTest.java
package ru.tinkoff.kora.json.annotation.processor;

import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.json.common.ListJsonReader;
import ru.tinkoff.kora.json.common.ListJsonWriter;
import tools.jackson.core.JsonGenerator;
import tools.jackson.core.JsonParser;

import java.util.List;

public class JsonIncludeTest extends AbstractJsonAnnotationProcessorTest {
    @Test
    public void testIncludeAlways() {
        compile("""
            import ru.tinkoff.kora.json.common.annotation.JsonInclude;

            @JsonInclude(JsonInclude.IncludeType.ALWAYS)
            @Json
            public record TestRecord(@Nullable Integer value) { }
            """
        );

        compileResult.assertSuccess();

        var mapper = mapper("TestRecord");
        mapper.verifyWrite(newObject("TestRecord", 42), "{\"value\":42}");
        mapper.verifyWrite(newObject("TestRecord", new Object[]{null}), "{\"value\":null}");
    }

    @Test
    public void testIncludeNonNull() {
        compile("""
            import ru.tinkoff.kora.json.common.annotation.JsonInclude;

            @JsonInclude(JsonInclude.IncludeType.NON_NULL)
            @Json
            public record TestRecord(@Nullable Integer value) { }
            """
        );

        compileResult.assertSuccess();

        var mapper = mapper("TestRecord");
        mapper.verifyWrite(newObject("TestRecord", 42), "{\"value\":42}");
        mapper.verifyWrite(newObject("TestRecord", new Object[]{null}), "{}");
    }

    @Test
    public void testIncludeNonEmpty() {
        compile("""
            import ru.tinkoff.kora.json.common.annotation.JsonInclude;

            @JsonInclude(JsonInclude.IncludeType.NON_EMPTY)
            @Json
            public record TestRecord(java.util.@Nullable List<Integer> value) { }
            """
        );

        compileResult.assertSuccess();

        var mapper = mapper("TestRecord", List.of(new ListJsonReader<>(JsonParser::getIntValue)), List.of(new ListJsonWriter<Integer>(JsonGenerator::writeNumber)));
        mapper.verifyWrite(newObject("TestRecord", List.of(42)), "{\"value\":[42]}");
        mapper.verifyWrite(newObject("TestRecord", List.of()), "{}");
        mapper.verifyWrite(newObject("TestRecord", new Object[]{null}), "{}");
    }

    @Test
    public void testFieldIncludeAlways() {
        compile("""
            import ru.tinkoff.kora.json.common.annotation.JsonInclude;

            @Json
            public record TestRecord(@Nullable String name, @JsonInclude(JsonInclude.IncludeType.ALWAYS) @Nullable Integer value) { }
            """
        );

        compileResult.assertSuccess();

        var mapper = mapper("TestRecord");
        mapper.verifyWrite(newObject("TestRecord", "test", 42), "{\"name\":\"test\",\"value\":42}");
        mapper.verifyWrite(newObject("TestRecord", null, null), "{\"value\":null}");
    }

    @Test
    public void testFieldIncludeNonEmpty() {
        compile("""
            import ru.tinkoff.kora.json.common.annotation.JsonInclude;

            @Json
            public record TestRecord(@JsonInclude(JsonInclude.IncludeType.NON_EMPTY) java.util.@Nullable List<Integer> value) { }
            """
        );

        compileResult.assertSuccess();

        var mapper = mapper("TestRecord", List.of(new ListJsonReader<>(JsonParser::getIntValue)), List.of(new ListJsonWriter<Integer>(JsonGenerator::writeNumber)));
        mapper.verifyWrite(newObject("TestRecord", List.of(42)), "{\"value\":[42]}");
        mapper.verifyWrite(newObject("TestRecord", List.of()), "{}");
        mapper.verifyWrite(newObject("TestRecord", new Object[]{null}), "{}");
    }
}

File: json/json-symbol-processor/src/test/kotlin/io/koraframework/json/ksp/JsonIncludeTest.kt
package io.koraframework.json.ksp

import org.junit.jupiter.api.Test
import io.koraframework.json.common.ListJsonReader
import io.koraframework.json.common.ListJsonWriter
import tools.jackson.core.JsonGenerator
import tools.jackson.core.JsonParser

class JsonIncludeTest : AbstractJsonSymbolProcessorTest() {
    @Test
    fun testIncludeAlways() {
        compile("""
            import io.koraframework.json.common.annotation.JsonInclude
                            
            @JsonInclude(JsonInclude.IncludeType.ALWAYS)
            @Json
            data class TestRecord(val value: Int?)
            """
            .trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("TestRecord")
        mapper.assertWrite(new("TestRecord", 42), "{\"value\":42}")
        mapper.assertWrite(new("TestRecord", null), "{\"value\":null}")
    }

    @Test
    fun testIncludeNonNull() {
        compile("""
            import io.koraframework.json.common.annotation.JsonInclude;
                            
            @JsonInclude(JsonInclude.IncludeType.NON_NULL)
            @Json
            data class TestRecord(val value: Int?)
            """
            .trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("TestRecord")
        mapper.assertWrite(new("TestRecord", 42), "{\"value\":42}")
        mapper.assertWrite(new("TestRecord", null), "{}")
    }

    @Test
    fun testIncludeNonEmpty() {
        compile("""
            import io.koraframework.json.common.annotation.JsonInclude;
                            
            @JsonInclude(JsonInclude.IncludeType.NON_EMPTY)
            @Json
            data class TestRecord(val value: List<Int>?)
            
            """
            .trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("TestRecord", listOf(ListJsonReader<Int> { obj: JsonParser -> obj.intValue }), listOf(ListJsonWriter<Int> { obj: JsonGenerator, v: Int? -> obj.writeNumber(v!!) }))
        mapper.assertWrite(new("TestRecord", listOf(42)), "{\"value\":[42]}")
        mapper.assertWrite(new("TestRecord", listOf<Any>()), "{}")
        mapper.assertWrite(new("TestRecord", null), "{}")
    }

    @Test
    fun testFieldIncludeAlways() {
        compile("""
            import io.koraframework.json.common.annotation.JsonInclude;
                            
            @Json
            data class TestRecord(val name: String?, @JsonInclude(JsonInclude.IncludeType.ALWAYS) val value: Int?)
            """
            .trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("TestRecord")
        mapper.assertWrite(new("TestRecord", "test", 42), "{\"name\":\"test\",\"value\":42}")
        mapper.assertWrite(new("TestRecord", null, null), "{\"value\":null}")
    }

    @Test
    fun testFieldIncludeNonEmpty() {
        compile("""
            import io.koraframework.json.common.annotation.JsonInclude;
                            
            @Json
            data class TestRecord(@JsonInclude(JsonInclude.IncludeType.NON_EMPTY) val value: List<Int>?)
            """
            .trimIndent()
        )
        compileResult.assertSuccess()
        val mapper = mapper("TestRecord", listOf(ListJsonReader<Int> { obj: JsonParser -> obj.intValue }), listOf(ListJsonWriter<Int> { obj: JsonGenerator, v: Int? -> obj.writeNumber(v!!) }))
        mapper.assertWrite(new("TestRecord", listOf(42)), "{\"value\":[42]}")
        mapper.assertWrite(new("TestRecord", listOf<Any>()), "{}")
        mapper.assertWrite(new("TestRecord", null), "{}")
    }
}
